core/src/main/java/dk/alexandra/fresco/lib/collections/permute/WaksmanUtils.java

```diff
@@ -0,0 +1,212 @@
+    }
+
+    int[] I = new int[n / 2];
+    int[] O = new int[n / 2];
+    for (int j = 0; j < O.length; j++) {
+      I[j] = -1;
+      O[j] = -1;
+    }
+
+    int[] topPerm = new int[n / 2];
+    int[] bottomPerm = new int[n / 2];
+
+    int[] permInv = invert(perm);
+
+    while (true) {
+
+      // find next O swapper that doesn't have control bit set
+      int unsetOSwapperIdx = nextUnsetSwapperIdx(O);
+      // if there is no such swapper, we have set all I and O swappers
+      // at this level of the network and can stop
+      if (unsetOSwapperIdx == -1) {
+        break;
+      }
+
+      int OSwIdx = unsetOSwapperIdx;
+      O[OSwIdx] = 0;
+
+      int firstOElIdx = topElIndexOf(OSwIdx);
+      int OElIdx = firstOElIdx;
+
+      while (true) {
+
+        /*
+         * we will "route" from the current output element to the input element that maps to it in
+         * the given permutation via the top permutation.
+         */
+
+        // look up which input element maps to this output element in
+        // the permutation
+        int IElIdx = permInv[OElIdx];
+
+        // get the swapper of the input element
+        int ISwIdx = swapperIdxOf(IElIdx);
+
+        // set top sub-permutation for the selected I and O swappers
+        topPerm[ISwIdx] = OSwIdx;
+
+        /*
+         * we are routing through the top permutation and by convention all top elements must go to
+         * the top permutation and all bottom elements must go to bottom permutation. if our input
+         * element is the bottom element we must therefore set the input swapper to "swap"
+         */
+        if (isTopEl(IElIdx)) {
+          I[ISwIdx] = 0;
+        } else {
+          I[ISwIdx] = 1;
+        }
+
+        /*
+         * now that we have set an I swapper, we can take the other input element associated with
+         * this swapper, find which output element it maps to in our permutation and set its
+         * corresponding O swapper by "routing" to it through the bottom permutation
+         */
+        IElIdx = neighElIdxOf(IElIdx);
+
+        OElIdx = perm[IElIdx];
+
+        // look up swapper of output element
+        OSwIdx = swapperIdxOf(OElIdx);
+
+        // we are routing through the bottom permutation so if output
+        // element
+        // is top we need to set swapper to "swap"
+        if (isTopEl(OElIdx)) {
+          O[OSwIdx] = 1;
+        } else {
+          O[OSwIdx] = 0;
+        }
+
+        // update bottom permutation
+        bottomPerm[ISwIdx] = OSwIdx;
+
+        // get neighbor element of output element
+        OElIdx = neighElIdxOf(OElIdx);
+
+        /*
+         * note that we are guaranteed to terminate since we are picking the next element to visit
+         * based on a permutation so the only element we are ever going to revisit is the one we
+         * started with at which point we stop
+         */
+        if (OElIdx == firstOElIdx) {
+          break;
+        }
+      }
+    }
+
+    /*
```

Comment:
There is a lot of comments in this method, which is hard to read - could we split it up in sub methods (named as stated in the comment) and get easier to read code?